File: clients/elixir/test/client_test.exs
defmodule VexillaClientTest do
  use ExUnit.Case
  # doctest VexillaClient

  test "fetches feature flags" do
    HTTPoison.start()

    user_id = "b7e91cc5-ec76-4ec3-9c1c-075032a13a1a"

    test_server_host = System.get_env("TEST_SERVER_HOST", "localhost:3000")

    config =
      VexillaClient.create_config(
        "http://#{test_server_host}",
        "dev",
        user_id
      )
      |> VexillaClient.sync_manifest!(&generic_fetch!/1)
      |> VexillaClient.sync_flags!("Gradual", &generic_fetch!/1)

    should_gradual = VexillaClient.should?(config, "Gradual", "testingWorkingGradual")
    assert should_gradual == true

    should_not_gradual = VexillaClient.should?(config, "Gradual", "testingNonWorkingGradual")
    assert should_not_gradual == false

    config = VexillaClient.sync_flags!(config, "Selective", &generic_fetch!/1)

    should_selective_string = VexillaClient.should?(config, "Selective", "String")
    assert should_selective_string == true

    should_selective_custom_string =
      VexillaClient.should_custom?(config, "Selective", "String", "shouldBeInList")

    assert should_selective_custom_string == true

    should_selective_number = VexillaClient.should_custom?(config, "Selective", "Number", 42)
    assert should_selective_number == true

    config = VexillaClient.sync_flags!(config, "Value", &generic_fetch!/1)

    value_string = VexillaClient.value!(config, "Value", "String", "bar")
    assert value_string == "foo"

    value_string_bad = VexillaClient.value!(config, "Value", "String_Bad", "bar")
    assert value_string_bad == "bar"

    value_int = VexillaClient.value!(config, "Value", "Integer", 0)
    assert value_int == 42

    value_int_bad = VexillaClient.value!(config, "Value", "Integer_Bad", 0)
    assert value_int_bad == 0

    value_float = VexillaClient.value!(config, "Value", "Float", 0)
    assert value_float == 42.42

    value_float_bad = VexillaClient.value!(config, "Value", "Float_Bad", 0)
    assert value_float_bad == 0

    config = VexillaClient.sync_flags!(config, "Scheduled", &generic_fetch!/1)

    should_before_global = VexillaClient.should?(config, "Scheduled", "beforeGlobal")
    assert should_before_global == false
    should_during_global = VexillaClient.should?(config, "Scheduled", "duringGlobal")
    assert should_during_global == true
    should_after_global = VexillaClient.should?(config, "Scheduled", "afterGlobal")
    assert should_after_global == false

    should_before_global_start_end =
      VexillaClient.should?(config, "Scheduled", "beforeGlobalStartEnd")

    assert should_before_global_start_end == false

    should_during_global_start_end =
      VexillaClient.should?(config, "Scheduled", "duringGlobalStartEnd")

    assert should_during_global_start_end == true

    should_after_global_start_end =
      VexillaClient.should?(config, "Scheduled", "afterGlobalStartEnd")

    assert should_after_global_start_end == false

    should_before_global_daily = VexillaClient.should?(config, "Scheduled", "beforeGlobalDaily")
    assert should_before_global_daily == false
    should_during_global_daily = VexillaClient.should?(config, "Scheduled", "duringGlobalDaily")
    assert should_during_global_daily == true
    should_after_global_daily = VexillaClient.should?(config, "Scheduled", "afterGlobalDaily")
    assert should_after_global_daily == false
  end

  # Fetch the JSON and decode it.
  defp generic_fetch!(url) do
    response = HTTPoison.get!(url)
    Jason.decode!(response.body)
  end
end

File: clients/kotlin/src/test/kotlin/ClientTest.kt
import org.junit.jupiter.api.Test
import kotlin.test.*
import dev.vexilla.Client
import dev.vexilla.GradualGroup
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.runBlocking
import java.nio.charset.Charset

class ClientTest {
    @Test
    fun `should handle testing against the test-server repo`() {

        val httpClient = HttpClient(CIO)

        val testServerHost: String = System.getenv("TEST_SERVER_HOST") ?: "http://localhost:3000"

        val uuid = "b7e91cc5-ec76-4ec3-9c1c-075032a13a1a"
        val vexillaClient = Client(testServerHost, "dev", uuid)

        runBlocking {
            vexillaClient.syncManifest { url ->
                httpClient.get(url).bodyAsText(Charset.defaultCharset())
            }

            vexillaClient.syncFlags(GradualGroup.name) { url ->
                httpClient.get(url).bodyAsText(Charset.defaultCharset())
            }

            vexillaClient.syncFlags("Selective") { url ->
                httpClient.get(url).bodyAsText(Charset.defaultCharset())
            }

            vexillaClient.syncFlags("Value") { url ->
                httpClient.get(url).bodyAsText(Charset.defaultCharset())
            }

            vexillaClient.syncFlags("Scheduled") { url ->
                httpClient.get(url).bodyAsText(Charset.defaultCharset())
            }
        }

        val shouldGradual = vexillaClient.should("Gradual", "testingWorkingGradual")
        assertTrue(shouldGradual.isSuccess)
        assertEquals(true, shouldGradual.getOrNull())

        val shouldNotGradual = vexillaClient.should("Gradual", "testingNonWorkingGradual")
        assertTrue(shouldNotGradual.isSuccess)
        assertEquals(false, shouldNotGradual.getOrNull())

        val shouldSelective = vexillaClient.should("Selective", "String")
        assertTrue(shouldSelective.isSuccess)
        assertEquals(true, shouldSelective.getOrNull())

        val shouldSelectiveCustomString = vexillaClient.should("Selective", "String", "shouldBeInList")
        assertTrue(shouldSelectiveCustomString.isSuccess)
        assertEquals(true, shouldSelectiveCustomString.getOrNull())

        val shouldNotSelectiveCustomString = vexillaClient.should("Selective", "String", "shouldNOTBeInList")
        assertTrue(shouldNotSelectiveCustomString.isSuccess)
        assertEquals(false, shouldNotSelectiveCustomString.getOrNull())

        val shouldSelectiveCustomInt = vexillaClient.shouldCustomInt("Selective", "Number", 42)
        assertTrue(shouldSelectiveCustomInt.isSuccess)
        assertEquals(true, shouldSelectiveCustomInt.getOrNull())

        val shouldNotSelectiveCustomInt = vexillaClient.shouldCustomInt("Selective", "Number", 43)
        assertTrue(shouldNotSelectiveCustomInt.isSuccess)
        assertEquals(false, shouldNotSelectiveCustomInt.getOrNull())

        val valueString = vexillaClient.valueString("Value", "String", "")
        assertTrue(valueString.isSuccess)
        assertEquals("foo", valueString.getOrNull())

        val valueInt = vexillaClient.valueInt("Value", "Integer", 0)
        assertTrue(valueInt.isSuccess)
        assertEquals(42, valueInt.getOrNull())

        val valueFloat = vexillaClient.valueFloat("Value", "Float", 0.0f)
        assertTrue(valueFloat.isSuccess)
        assertEquals(42.42f, valueFloat.getOrNull())
        

        val beforeGlobal = vexillaClient.should("Scheduled", "beforeGlobal")
        assertTrue(beforeGlobal.isSuccess)
        assertEquals(false, beforeGlobal.getOrNull())

        val duringGlobal = vexillaClient.should("Scheduled", "duringGlobal")
        assertTrue(duringGlobal.isSuccess)
        assertEquals(true, duringGlobal.getOrNull())

        val afterGlobal = vexillaClient.should("Scheduled", "afterGlobal")
        assertTrue(afterGlobal.isSuccess)
        assertEquals(false, afterGlobal.getOrNull())


        val beforeGlobalStartEnd = vexillaClient.should("Scheduled", "beforeGlobalStartEnd")
        assertTrue(beforeGlobalStartEnd.isSuccess)
        assertEquals(false, beforeGlobalStartEnd.getOrNull())

        val duringGlobalStartEnd = vexillaClient.should("Scheduled", "duringGlobalStartEnd")
        assertTrue(duringGlobalStartEnd.isSuccess)
        assertEquals(true, duringGlobalStartEnd.getOrNull())

        val afterGlobalStartEnd = vexillaClient.should("Scheduled", "afterGlobalStartEnd")
        assertTrue(afterGlobalStartEnd.isSuccess)
        assertEquals(false, afterGlobalStartEnd.getOrNull())


        val beforeGlobalDaily = vexillaClient.should("Scheduled", "beforeGlobalDaily")
        assertTrue(beforeGlobalDaily.isSuccess)
        assertEquals(false, beforeGlobalDaily.getOrNull())

        val duringGlobalDaily = vexillaClient.should("Scheduled", "duringGlobalDaily")
        assertTrue(duringGlobalDaily.isSuccess)
        assertEquals(true, duringGlobalDaily.getOrNull())

        val afterGlobalDaily = vexillaClient.should("Scheduled", "afterGlobalDaily")
        assertTrue(afterGlobalDaily.isSuccess)
        assertEquals(false, afterGlobalDaily.getOrNull())
    }
}